### ISSUE TITLE
update realtime solar/geomagnetic forcing input to use solar wind and ovation parameters

### ISSUE THREAD

**Issue by akubaryk** [SWPC-IPE/WAM-IPE#285]
Need to find where these files are on WCOSS and then write processing scripts to get them into the model for INPUT_PARAMETERS=realtime in the same way that the post-processed input files are parsed (INPUT_PARAMETERS=timeobs).

**Comment by akubaryk** [comment]
We're still waiting on DTS to provide us an update on making these realtime parameters available in production. Need to change the filename formatting so we can start building up a database of realtime parameters for solar wind/ovation, as well, as has been done with F10.7/Kp xml files.

**Comment by akubaryk** [comment]
Branch `realtime_ipe_refactor_merge` has been created as this code needs a review before it goes into #336. The changes discussed in this comment are visible in 778fa81.

This is the new logic for realtime input parameter parsing:

The Python script is passed a "wam_input_new" file from which F10.7, F10.7d, Kp, and 24hr Kp avg are taken. Those are interpolated to minute cadence, and then starting from the first date for those values, the script will begin to pull information out of the geospace-input xml files (in dcom) for solar wind parameters: Bz, By, SW Velocity, SW Density. From this Bt and SW Angle are calculated. These are already at minute cadence, which makes things simple.

Then, the script looks in the WAMINDIR for the swpc_aurora_power files: there is one for each date, and they are broken up into five minute chunks. The script rounds the "current time" it is searching for to the nearest 5 minutes, and then searches the file for the corresponding NH/SH gigawatts, and then derives the hemispheric power index from that. No additional interpolation is applied.

There are many cases where data is missing. In the event of missing geospace/SW data, the script simply derives the solar wind parameters in the same manner it used to. In the event that aurora power data is missing, it searches backwards in the file 5 minutes (from the rounded value), and then forward five minutes, and then falls back on deriving gigawatts from Kp if it still can't find a match. We could change the SW fallback to search forwards or backwards a minute (or more), as well.

A flat delay of 40 minutes is applied to the solar wind parameters (20 minutes from L1, 20 minutes for the magnetospheric response time), and a further 20 minutes of backward averaging is done to get the final Bz, Bt, and SW Angle values.

Comments are welcome prior to the pull request.

**Comment by wanghj2015** [comment]
@akubaryk I'll take a look at this. What are the path to the 'solar wind' and 'aurora power' data files?
Is 'aurora-nowcast-hemi-power.txt' now time stamped and saved, or still just one file and overwritten each day?

**Comment by akubaryk** [comment]
WAMINDIR contains the timestamped swpc_aurora_power files and the wam_input_new files, and the path to the geospace-input xml files is set by default in parse.py, take a look at the args setup for the default `-g` path.

**Comment by timfullerrowell** [comment]
Thanks Adam,

This is great. The missing data is unfortunate, but probably something we will have to live with for now in the real-time system. It's good you have fixes for most of the missing values for now. When you get a minute(!) we should give feedback to DTS to show them the issues we are facing. There will likely also be "bad" data in the real-time solar wind observation, that we might need traps for. We'll get input from Doug Biesecker when we start to questions some of the values.

One correction in the delays applied to the solar wind parameters. I think we need 30 minutes for the solar wind propagation time, rather than 20 mins, so add 50 mins plus the 20 minute averaging. I'll double check old emails, but I think that is what we decided.

Thanks again,
Tim

**Comment by akubaryk** [comment]
I'll push up 50 minutes instead of 40 now and think about drafting a simplified version of the issues to send to DTS.

**Comment by akubaryk** [comment]
Looks like it was already 50, guess my typo was here instead of there.

**Comment by wanghj2015** [comment]
I'll take a look at it in detail later. But do you search across days in 'sw' and 'power' data? The 'sw' and 'power' data are group in day, not as f107/kp data which contain data of +/- 3 days for each xml file. This would be useful for historical runs.

**Comment by akubaryk** [comment]
Yes, of course it searches across multiple days. The geospace files are one minute *per file*, which means it has to search across as many files as it needs. I'm not sure what you mean by historical runs, but the swpc_aurora_power data only stretches back to early August, and the solar wind geospace data is regularly wiped from dcom after just a couple weeks.

**Comment by wanghj2015** [comment]
A few comments:
1. The output for the _power_ in the old scripts is in the order '_NHemiPow NHemiPowIdx    SHemiPow SHemiPowIdx_', but that from the new scripts is in the order '_NHemiPow SHemiPow NHemiPowIdx SHemiPowIdx_'.
2. Should we use _interpolated_ data from each dataset (i.e., 'solar wind' and 'aurora') where they exist for the period (but just missing a few data points), instead of inserting them into the 'derived' data list? This can make the data list more consistent (within each dataset). 
3. When these 'sw' and 'aurora' data don't exist (i.e. for the future), should we use the last data point in the data list from 'sw' and 'aurora' (i.e., 'persistent'), or just switch to the 'derived' data, or even the blend of the two datasets? If we just use the 'derived' data, should we have a transition period (say 3 hr) from the 'sw' or 'aurora' data to the 'derived' data.

**Comment by wanghj2015** [comment]
@akubaryk If you wanted to test the scripts on 'hera', I have copied some 'sw' data here : 

/scratch1/NCEPDEV/swpc/sw

I have tested and compared the new and old scripts for '-s 2019092812 -d 24'.

**Comment by akubaryk** [comment]
@wanghj2015 has noted that wam_input_new files are occasionally missing the `<f107-41-avg>` tag for the majority of their 3hr chunks, so that will need to be addressed as well. I ran a quick test on all of the available wam_input_new files, and there are four bad days: 20180711, 20180712 (ends at 12Z), 20190128, and 20190802. Will likely wind up searching backwards through previous wam_input_new files, but we have several other options.

**Comment by akubaryk** [comment]
1 above (incorrect ordering) has been addressed with a recent commit into the realtime_ipe_refactor_merge branch, and so this branch should actually function now. 2 and 3 need to be decided on as a group; this is a topic for a future meeting. I'm not going to bother with the gaps in the f107-41-avg until we have a comprehensive plan for how to address missing realtime data.
